File: Day322_2025/Q322_GFG.cpp
class Solution {
  public:
    int countPaths(int V, vector<vector<int>>& edges) {
        // Q. Number of Ways to Arrive at Destination
        // code here
        int MOD = 1e9 + 7;
        
        vector<vector<pair<int, int>>> adj(V);
        for (int i = 0; i < edges.size(); i++) {
            adj[edges[i][0]].push_back({edges[i][1], edges[i][2]});
            adj[edges[i][1]].push_back({edges[i][0], edges[i][2]});
        }
        
        vector<long long> dist (V, LLONG_MAX);
        vector<int> ways(V, 0);
        dist[0] = 0;
        ways[0] = 1;
        
        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;
        pq.push({0, 0});
        
        while(!pq.empty()) {
            long long distance = pq.top().first;
            int node = pq.top().second;
            
            pq.pop();
            
            for (auto nbr: adj[node]) {
                int adjNode = nbr.first;
                int wt = nbr.second;
                
                if (distance + wt < dist[adjNode]) {
                    dist[adjNode] = distance + wt;
                    pq.push({dist[adjNode], adjNode});
                    ways[adjNode] = ways[node];
                }
                else if (distance + wt == dist[adjNode]) {
                    ways[adjNode] = (ways[adjNode] + ways[node]) % MOD;
                }
            }
        }
        return ways[V-1];
    }
};

File: Day322_2025/Q322_GFG.py
import heapq

class Solution:
    def countPaths(self, V, edges):
        MOD = 10**9 + 7
        
        adj = [[] for _ in range(V)]
        for u, v, w in edges:
            adj[u].append((v, w))
            adj[v].append((u, w))
        
        dist = [float('inf')] * V
        ways = [0] * V
        dist[0] = 0
        ways[0] = 1
        
        pq = [(0, 0)]  # (distance, node)
        
        while pq:
            currDist, node = heapq.heappop(pq)
            
            for adjNode, wt in adj[node]:
                if currDist + wt < dist[adjNode]:
                    dist[adjNode] = currDist + wt
                    ways[adjNode] = ways[node]
                    heapq.heappush(pq, (dist[adjNode], adjNode))
                
                elif currDist + wt == dist[adjNode]:
                    ways[adjNode] = (ways[adjNode] + ways[node]) % MOD
        
        return ways[V - 1]
